---
title: "homer motif/chip enrichment"
output: html_notebook
---

```{r bash}
#i run homer on the cluster because the environment already has some dependencies downloaded
#http://homer.ucsd.edu/homer/introduction/install.html

#download conda 
#set up virtual environment for isoseq3
ml fhPython/3.8.6-foss-2020b-Python-3.8.6 
ml Anaconda3/2022.05
ml SAMtools/1.11-GCC-10.2.0

#make conda environment named py3
conda create -n py3 python=3

grabnode
30
600
1
n

source activate py3

conda init zsh
conda install -c bioconda
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge

conda install wget samtools r-essentials bioconductor-deseq2 bioconductor-edger

#make new folder for homer, i like to put it in my ~/dev folder

#follow the rest of the tutorial
```

HOMER will be installed in the same directory that you place the configureHomer.pl program.  configureHomer.pl will attempt to check for required utilities and alert you to missing programs.
For the latest version of Homer, go to http://homer.salk.edu/homer/.
Download the "configureHomer.pl" script and place it in a directory where you would like homer to be installed (i.e. /home/chucknorris/homer/).
Run the configureHomer.pl script to install homer.
i.e. perl /Users/chucknorris/homer/configureHomer.pl -install
Add the homer/bin directory to your executable path.  For example, edit your ~/.bash_profile file to include the line:
PATH=$PATH:/Users/chucknorris/homer/bin/
NOTE: If ~/.bash_profile doesn't exist in your mac, create a new file in your home directory, place the PATH=... line in it, and them rename the file using the command-line prompt: "mv newfilename ~/.bash_profile"
Reset your terminal so that the changes to the PATH variable take effect
source ~/.bash_profile
You should now be able to execute programs in the homer/bin directory by just typing their name.

```{bash}
perl /path-to-homer/configureHomer.pl -install hg38

cd bin
annotatePeaks.pl #check to see if this works
```

```{bash}
#now that it's downlaoded, when you use it again this is how do configure

cd ~/dev/homer

grabnode
#get some cores!!

perl configureHomer.pl -make

cd bin
```

```{r prepare peaks as beds for homer}
library(Signac)
library(GenomicRanges)
link_df<- read.csv("~/m2/res/3_subgroup_links.csv")
link_df<-link_df[!duplicated(link_df$peak),]
k1<- link_df[link_df$kmean == "1",]
k2<- link_df[link_df$kmean == "2",]
k3<- link_df[link_df$kmean == "3",]

peaks<- StringToGRanges(k2$peak, sep = c(":", "-", "-"))
values(peaks)<- DataFrame(name = k2$peak)
peaks

# Column1: chromosome
# Column2: starting position
# Column3: ending position
# Column4: Unique Peak ID
# Column5: not used
# Column6: Strand (+/- or 0/1, where 0="+", 1="-")

df <- data.frame(seqnames=seqnames(peaks),
  starts=start(peaks)-1,
  ends=end(peaks),
  names=values(peaks),
  scores=c(rep(".", length(peaks))),
  strands="+")

write.table(df, file="~/m2/k2_peaks.bed", quote=F, sep="\t", row.names=F, col.names=F)

known_homer<-read.delim("~/m2/res/k3_homer/knownResults.txt")
```

```{bash}
findMotifsGenome.pl ~/m2/k1_peaks.bed hg38 ~/m2/res -size 500 -p 12
```

```{r homer results}
k3_homer<-read.delim("~/m2/res/k3_homer/knownResults.txt")
k2_homer<-read.delim("~/m2/res/k2_homer/knownResults.txt")
```

```{bash kmean peaks centered at ewsfli motif}
annotatePeaks.pl ~/m2/k3_peaks.bed hg38 -size 500 -center ~/m2/res/motif105.motif ~/m2/res/ap1.motif ~/m2/res/ap1.motif ~/m2/res/fosl2.motif > ~/m2/res/ewsfli_k3_CenteredPeaks.txt

annotatePeaks.pl ~/m2/k1_peaks.bed hg38 -size 500 -center ~/m2/res/motif105.motif ~/m2/res/ap1.motif ~/m2/res/ap1.motif ~/m2/res/fosl2.motif > ~/m2/res/ewsfli_k1_CenteredPeaks.txt
```

```{bash kmean peaks, motif enrichment of top hits (from results)}
annotatePeaks.pl ~/m2/k1_peaks.bed hg38 -size 500 -hist 5 -m ~/m2/res/motifs/*.motif > ~/m2/res/motif_enrich/motif_k1.txt

annotatePeaks.pl ~/m2/k2_peaks.bed hg38 -size 500 -hist 5 -m ~/m2/res/motifs/*.motif > ~/m2/res/motif_enrich/motif_k2.txt

annotatePeaks.pl ~/m2/k3_peaks.bed hg38 -size 500 -hist 5 -m ~/m2/res/motifs/*.motif > ~/m2/res/motif_enrich/motif_k3.txt
```

```{r visualize motif enrichment}
kmeans<-c("k1", "k2", "k3")

x<-"k1"

mtf_list<-lapply(kmeans, function(i){
  
  mtf<-read.delim(paste0("~/m2/res/motif_enrich/motif_",i, ".txt"))
  avg_mtf<-lapply(seq(from=2,to=length(names(mtf))-4,by=3), function(x){
    df<-data.frame(distance =  mtf[,1], mtf = rowMeans( mtf[,x:(x+2)]), motif =  str_split(names(mtf)[x], "\\.\\.") %>%  sapply("[[", 1))
    df
  })
  
  df<-do.call(rbind, avg_mtf)
  df$kmean <- i
  df
})

mtf_df<-do.call(rbind, mtf_list)

table(mtf_df$motif)

#ews plot
pdf("~/m2/Plots/kmean_ews.pdf", width = 8, height = 3)
ews_mtf<-mtf_df[mtf_df$motif %in% c("EWS.ERG.fusion.ETS", "EWS.FLI1.fusion.ETS",  "Fli1.ETS"),]
ggplot(ews_mtf, aes(x = distance, y = mtf, color = motif))+geom_smooth()+theme_classic()+scale_color_manual(values = brewer.greys(n=9)[c(3,6,9)])+facet_wrap(~kmean)
dev.off()

#k1 plot
k1_mtf<-mtf_df[mtf_df$motif %in% c("Jun.AP1.bZIP", "Sp1.Zf", "X2.TCCGATTGGC.BestGuess.NFY.CCAAT", "NeuroD1.bHLH", "Fos.bZIP"),]
k1_mtf$motif[k1_mtf$motif == "X2.TCCGATTGGC.BestGuess.NFY.CCAAT"]<-"NFY"
pdf("~/m2/Plots/k1_homer.pdf", width = 8, height = 3)
ggplot(k1_mtf, aes(x = distance, y = mtf, color = motif))+geom_smooth()+theme_classic()+scale_color_manual(values = brewer.bupu(n=5))+facet_wrap(~kmean)
dev.off()
#k2 plot
k2_mtf<-mtf_df[mtf_df$motif %in% c("TEAD1.TEAD", "Fosl2.bZIP", "Hoxd13.Homeobox", "Smad2.MAD"),]
pdf("~/m2/Plots/k2_homer.pdf", width = 8, height = 3)
ggplot(k2_mtf, aes(x = distance, y = mtf, color = motif))+geom_smooth()+theme_classic()+scale_color_manual(values = brewer.paired(n = 4))+facet_wrap(~kmean)
dev.off()

#k3 plot
k3_mtf<-mtf_df[mtf_df$motif %in% c( "GABPA.ETS", "NRF1.NRF", "KLF5.Zf"),]
pdf("~/m2/Plots/k3_homer.pdf", width = 8, height = 3)
ggplot(k3_mtf, aes(x = distance, y = mtf, color = motif))+geom_smooth()+theme_classic()+scale_color_manual(values = brewer.paired(n =9)[6:9])+facet_wrap(~kmean)
dev.off()
```
#-------------------------------------------------------------------------------------------------------------------#
```{r prepare cell line fli binding beds}
peak_dir<-"/fh/fast/furlan_s/user/owalt/ewings/fli_chip/macs2"
cl<-c("A673","RDES", "SKNMC", "CHLA10", "TC71", )
x<-"RDES"
peakl<-lapply(cl, function(x){
  df<-read.table(file.path(peak_dir, paste0(x, "_fli_chip_peaks_peaks.narrowPeak")))  
  colnames(df)<-c("chr start end name integer_score none fold_enrichment log10pvalue log10qvalue relative_summit_pos") %>% strsplit(" ") %>% unlist()
  gr<-makeGRangesFromDataFrame(df)
  
  peakwidths <- width(gr)
  gr <- gr[peakwidths  < 10000 & peakwidths > 20]
  gr <- keepStandardChromosomes(gr, pruning.mode = "coarse")
  
  gr$cell_line <-x
  gr
})  

#make beds
for(peaks in peakl){

  df <- data.frame(seqnames=seqnames(peaks),
                   starts=start(peaks)-1,
                   ends=end(peaks),
                   names=values(peaks),
                   scores=c(rep(".", length(peaks))),
                   strands="+")
  
  write.table(df, file=paste0("~/m2/res/",peaks$cell_line %>% unique(), ".bed"), quote=F, sep="\t", row.names=F, col.names=F)
  
}
```

```{bash kmean peaks, enrichment of fli binding sites}
annotatePeaks.pl ~/m2/k1_peaks.bed hg38 -size 500 -hist 5 -p ~/m2/res/cell_line_fli/*.bed > ~/m2/res/cell_line_fli/fli_k1.txt

annotatePeaks.pl ~/m2/k2_peaks.bed hg38 -size 500 -hist 5 -p ~/m2/res/cell_line_fli/*.bed > ~/m2/res/cell_line_fli/fli_k2.txt

annotatePeaks.pl ~/m2/k3_peaks.bed hg38 -size 500 -hist 5 -p ~/m2/res/cell_line_fli/*.bed > ~/m2/res/cell_line_fli/fli_k3.txt
```

```{r visualize fli binding site enrichment}
kmeans<-c("k1", "k2", "k3")
x<-"k1"
fli_list<-lapply(kmeans, function(x){
  fli<-read.delim(paste0("~/m2/res/cell_line_fli/fli_",x, ".txt"), sep = "\t")
  fli$kmean<-x
  names(fli)[1:6]<-c("distance","A673", "CHLA10","RDES","SKNMC", "TC71")
  fli
  
  fli_df<-lapply(names(fli)[2:6], function(x){
    df<-data.frame(distance =  fli[,1], enrich = fli[,x], cell_line =  x)
    df
  })
  
  df<-do.call(rbind, fli_df)
  df$kmean <- x
  df
  
  
})

fli_df<-do.call(rbind, fli_list)
pdf("~/m2/Plots/cl_fli_bind.pdf", width = 8, height = 3)
ggplot(fli_df, aes(x = distance, y = enrich, color = cell_line))+geom_smooth()+theme_classic()+scale_color_manual(values = c("A673" = "#D51F26",  "CHLA10" = "#89288F", "SKNMC"= "#8A9FD1", "TC71" = "#e9c2db","RDES" = "#f5e358"))+ggtitle(label = "Fli1 Binding Site Enrichment")+facet_wrap(~kmean)
dev.off()
```